### ISSUE TITLE
[java] UnusedLocalVariable false positive when variable is read as side effect of an assignment

### ISSUE THREAD

**Issue by aaime** [pmd/pmd#5061]
**Affects PMD Version:** 7.2

**Rule:**

[UnusedLocalVariable](https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#unusedlocalvariable)

**Description:**

A variable used in a read loop to check how many bytes are getting read, in a situation where it's both updated, but as a side effect of the assignment, also read.

**Code Sample demonstrating the issue:**

```java
      int read = 0;
      while ((read += channel.read(buffer)) < length) {
      }
```

**Expected outcome:**

PMD should not report the variable as unused.

**Running PMD through:** *Maven*

**Comment by aaime** [comment]
By the way, this code will also trigger an [EmptyControlStatement](https://pmd.github.io/pmd/pmd_rules_java_codestyle.html#emptycontrolstatement), but the desired action is happening in the loop condition itself.

**Comment by aaime** [comment]
Another case triggering the unused local variable is the following on ``lock``:

```java
try (FileOutputStream fos = new FileOutputStream(file);
                FileChannel out = fos.getChannel();
                FileLock lock = out.lock()) {
            final int buffsize = 4096;
            long position = 0;
            long read;
            while ((read = out.transferFrom(channel, position, buffsize)) > 0) {
                position += read;
            }
            return position;
        }
```

FileLock is AutoCloseable and the close will invoke release, so this is all the code that's needed.

**Comment by adangel** [comment]
About FileLock and unused variable `lock`: this is a different issue, see #5747 - and won't be fixed with this issue. You could rename the variable to "ignored".

Your code sample most likely has an error (at least, it makes some assumptions...): The [read](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/nio/channels/ReadableByteChannel.html#read(java.nio.ByteBuffer)) method would return "-1" if EOF is reached, which would be added up to the total "read" and your count is wrong in the end.

That's why such a loop is usually written like this:

```java
int totalBytesRead = 0;
int read = 0;
while ((read = channel.read(buffer)) >= 0 && totalBytesRead < length) {
    totalBytesRead += read;
}
```
